diff --git a/ParaViewCore/VTKExtensions/Rendering/vtkPVGeometryFilter.cxx b/ParaViewCore/VTKExtensions/Rendering/vtkPVGeometryFilter.cxx
index 15b6a7a..45a273a 100644
--- a/ParaViewCore/VTKExtensions/Rendering/vtkPVGeometryFilter.cxx
+++ b/ParaViewCore/VTKExtensions/Rendering/vtkPVGeometryFilter.cxx
@@ -1310,7 +1310,7 @@ void vtkPVGeometryFilter::StructuredGridExecute(vtkStructuredGrid* input, vtkPol
     {
       if (input->HasAnyBlankCells())
       {
-        this->DataSetSurfaceFilter->DataSetExecute(input, output);
+        this->DataSetSurfaceFilter->StructuredWithBlankingExecute(input, output);
       }
       else
       {
