diff --git a/ParaViewCore/ClientServerCore/Rendering/vtkUnstructuredGridVolumeRepresentation.cxx b/ParaViewCore/ClientServerCore/Rendering/vtkUnstructuredGridVolumeRepresentation.cxx
index d1d5ef278f..c208bdad79 100644
--- a/ParaViewCore/ClientServerCore/Rendering/vtkUnstructuredGridVolumeRepresentation.cxx
+++ b/ParaViewCore/ClientServerCore/Rendering/vtkUnstructuredGridVolumeRepresentation.cxx
@@ -330,7 +330,15 @@ void vtkUnstructuredGridVolumeRepresentation::UpdateMapperParameters()
     info->Has(vtkDataObject::FIELD_NAME()))
   {
     colorArrayName = info->Get(vtkDataObject::FIELD_NAME());
-    fieldAssociation = info->Get(vtkDataObject::FIELD_ASSOCIATION());
+    // The Resample To Image filter transforms cell data to point data.
+    if (this->Internals->ActiveVolumeMapper == "Resample To Image")
+    {
+      fieldAssociation = vtkDataObject::FIELD_ASSOCIATION_POINTS;
+    }
+    else
+    {
+      fieldAssociation = info->Get(vtkDataObject::FIELD_ASSOCIATION());
+    }
   }
 
   activeMapper->SelectScalarArray(colorArrayName);
