diff --git a/ParaViewCore/ClientServerCore/Core/vtkPVPlugin.h b/ParaViewCore/ClientServerCore/Core/vtkPVPlugin.h
index 12bfc3a..8c9c182 100644
--- a/ParaViewCore/ClientServerCore/Core/vtkPVPlugin.h
+++ b/ParaViewCore/ClientServerCore/Core/vtkPVPlugin.h
@@ -34,11 +34,10 @@
 #ifdef _WIN32
 // __cdecl gives an unmangled name
 # define C_DECL __cdecl
-# define C_EXPORT extern "C" __declspec(dllexport)
 #else
 # define C_DECL
-# define C_EXPORT extern "C"
 #endif
+#define C_EXPORT extern "C" VTKPVCLIENTSERVERCORECORE_EXPORT
 
 class VTKPVCLIENTSERVERCORECORE_EXPORT vtkPVPlugin
 {
