diff --git a/ParaViewCore/ServerManager/Rendering/vtkSMTransferFunctionProxy.cxx b/ParaViewCore/ServerManager/Rendering/vtkSMTransferFunctionProxy.cxx
index b3269dd..91ee87c 100644
--- a/ParaViewCore/ServerManager/Rendering/vtkSMTransferFunctionProxy.cxx
+++ b/ParaViewCore/ServerManager/Rendering/vtkSMTransferFunctionProxy.cxx
@@ -38,7 +38,8 @@
 #include "vtk_jsoncpp.h"
 
 #include <algorithm>
-#include <math.h>
+#include <cmath>
+#include <sstream>
 #include <vector>
 
 namespace
