diff --git a/ParaViewCore/ServerManager/Core/vtkSMSessionProxyManagerInternals.h b/ParaViewCore/ServerManager/Core/vtkSMSessionProxyManagerInternals.h
index e9e6339..2b79155 100644
--- a/ParaViewCore/ServerManager/Core/vtkSMSessionProxyManagerInternals.h
+++ b/ParaViewCore/ServerManager/Core/vtkSMSessionProxyManagerInternals.h
@@ -44,7 +44,7 @@
 class vtkSMProxyManagerProxyInfo : public vtkObjectBase
 {
 public:
-  vtkTypeMacro(vtkSMProxyManagerProxyInfo, vtkObjectBase);
+  vtkBaseTypeMacro(vtkSMProxyManagerProxyInfo, vtkObjectBase);
 
   vtkSmartPointer<vtkSMProxy> Proxy;
   unsigned long ModifiedObserverTag;
