diff --git a/ThirdParty/jsoncpp/CMakeLists.txt b/ThirdParty/jsoncpp/CMakeLists.txt
index a892e8c..c4c816f 100644
--- a/ThirdParty/jsoncpp/CMakeLists.txt
+++ b/ThirdParty/jsoncpp/CMakeLists.txt
@@ -1,2 +1,6 @@
 vtk_module_third_party(JsonCpp
   VERSION 0.7.0)
+if (WIN32 AND VTK_USE_SYSTEM_JSONCPP)
+  option(VTK_SYSTEM_JSONCPP_SHARED "System-installed JsonCpp package is a shared library" OFF)
+  mark_as_advanced(VTK_SYSTEM_JSONCPP_SHARED)
+endif()
\ No newline at end of file
diff --git a/ThirdParty/jsoncpp/vtk_jsoncpp.h.in b/ThirdParty/jsoncpp/vtk_jsoncpp.h.in
index ac6fd52..4c94ad1 100644
--- a/ThirdParty/jsoncpp/vtk_jsoncpp.h.in
+++ b/ThirdParty/jsoncpp/vtk_jsoncpp.h.in
@@ -18,13 +18,18 @@
 /* Use the JsonCpp library configured for VTK.  */
 #cmakedefine VTK_USE_SYSTEM_JSONCPP
 #ifdef VTK_USE_SYSTEM_JSONCPP
+#cmakedefine VTK_SYSTEM_JSONCPP_SHARED
+// Needed for Windows declspec import logic
+# if !defined(JSON_DLL) && defined(_WIN32) && defined(VTK_SYSTEM_JSONCPP_SHARED)
+#  define JSON_DLL
+# endif
 # include <json/json.h>
 # ifndef JSON_FORWARDS_H_INCLUDED
 #  error Incorrect header included, check your include paths for conflicts
 # endif
 #else
 // Needed for Windows declspec import logic
-# if defined(_WIN32) && defined(VTK_BUILD_SHARED_LIBS)
+# if !defined(JSON_DLL) && defined(_WIN32) && defined(VTK_BUILD_SHARED_LIBS))
 #  define JSON_DLL
 # endif
 # include <vtkjsoncpp/json/json.h>
