diff --git a/ThirdParty/netcdf/vtknetcdf/CMakeLists.txt b/ThirdParty/netcdf/vtknetcdf/CMakeLists.txt
index 4ed8168..93b1ac3 100644
--- a/ThirdParty/netcdf/vtknetcdf/CMakeLists.txt
+++ b/ThirdParty/netcdf/vtknetcdf/CMakeLists.txt
@@ -233,7 +233,7 @@ set (USE_NETCDF4 OFF)
 if (NOT VTK_USE_SYSTEM_HDF5)
   # using VTK's HDF5, we always build that with HL support.
   set (USE_NETCDF4 ON)
-elseif(HDF5_HL_LIBRARY OR HDF5_hdf5_hl_LIBRARY)
+elseif(HDF5_HL_LIBRARY OR HDF5_hdf5_hl_LIBRARY OR HDF5_hdf5_hl_LIBRARY_DEBUG OR HDF5_hdf5_hl_LIBRARY_RELEASE)
   set (USE_NETCDF4 ON)
 else()
   message(STATUS "Disabling NETCDF4 support since HDF5_HL or HDF5_hdf5_hl is missing.")
